### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [6]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [7]:
#Describe provides the list of stats:
#purchase_data.describe()
#Counts the number of unique rows in a column (playerct)
purchase_df = pd.DataFrame(purchase_data, columns=['Purchase ID', 'SN', 'Age', 'Gender', 'Item ID', 'Item Name', 'Price'])
players_counts = purchase_df['SN'].nunique()
players_df = pd.DataFrame([players_counts], columns=['Total Players'])
players_df

#purchase_data.count()
#purchase_data["Purchase ID"].value_counts()


Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [8]:
#1. Unique Items (unique_tot)
#purchase_data["Item ID"].unique()
unique_tot = purchase_df["Item ID"].max()
#pur_data1["Number of Unique Items"] = purchase_data["Item ID"].max()
#2. Average Price (average_price)
average_price = purchase_df["Price"].mean()
average_price_format = f'${average_price:.2f}'
#3. Number of Purchases (purchase_tot)
purchase_tot = purchase_df["Purchase ID"].max()

#4. Total Revenue (rev_tot)
rev_tot = purchase_df["Price"].sum()
#rev_tot_format = f'${average_price:.2f}'

unique_df = pd.DataFrame({"Number of Unique Items":[(unique_tot)], "Average Price":[(average_price_format)], 
                         "Number of Purchases":[(purchase_tot)], "Total Revenue":[(rev_tot)]})
unique_df["Total Revenue"] = unique_df["Total Revenue"].map("${:,.2f}".format)
unique_df.head()
#[unique_tot], [average_price], [purchase_tot], [rev_tot]
#purchase_data["SN"].value_counts()
#purchase_unique=["SN"].unique()
#purchase_unique()
#purchase_data.dtypes
#  Create a new Dataframe for the Count Variables?
#panaly_df = pd.Dataframe({"Number of Unique Lines":(unique_tot), 
#                             "Average Price":(average_price),
#                             "Number of Purchases": (purchase_tot),
#                             "Total Revenue": (rev_tot)}) 

#Sort the dataframe to determine the unique highest value.
#sort_purchase = purchase_data.sort_values("Item ID")
#sort_purchase["Item ID"].unique()

# Reference for formatting the values:
# First convert "average_donation", "goal", and "pledged" columns to float
# Then Format to go to two decimal places, include a dollar sign, and use comma notation
#hosted_in_us["average_donation"] = hosted_in_us["average_donation"].astype(float).map(
#    "${:,.2f}".format)
#hosted_in_us["goal"] = hosted_in_us["goal"].astype(float).map("${:,.2f}".format)
#hosted_in_us["pledged"] = hosted_in_us["pledged"].astype(float).map("${:,.2f}".format)

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  779     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [9]:
#Total Count by Gender Group
gender_groups = purchase_data.groupby("Gender")
gender_df = gender_groups['SN'].nunique()
#gender_df.rename(columns={"SN":"Total Counts"})
gender2_df = gender_df.to_frame().reset_index()
#gender2_df.head()
#gender2_df.rename(columns={"SN":"Total Counts"})

gensum = gender2_df["SN"].sum()

gender3 =  gender2_df["SN"] / gensum
#gender3

gender2_df["Percentage of Players"] = (gender2_df["SN"] / gensum) *100
gender2_sort_df = gender2_df.sort_values("Percentage of Players", ascending = False)
gender2_sort_df["Percentage of Players"] = gender2_sort_df["Percentage of Players"].map("{:.2f}".format)
gender2_sort_df = gender2_sort_df.rename(columns={"SN":"Total Counts"})
gender2_sort_df.head()

#Getting value counts to %
#s = pd.Series([3, 1, 2, 3, 4, np.nan])
#s.value_counts(normalize=True)

#gender_groups
#gender_groups.head()

#Average calculation is the gender groups/(playerct)


Gender  Total Counts Percentage of Players
1                   Male           484                 84.03
0                 Female            81                 14.06
2  Other / Non-Disclosed            11                  1.91


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [10]:
#gen_df = purchase_data.set_index("Gender")
#gen_df.head()
#pokemon_group = pokemon_type.groupby(["Type 1"])
#gender_type = purchase_data[["Gender", "Purchase ID", "Item ID", "Price"]]
#gen2_df = purchase_data.set_index("Gender")
#gen2_df.head()
#gen2_df.dtype()
gender_counts = purchase_data["Gender"].value_counts()
#gender_counts.head()
gender2groups_df = purchase_data.groupby(["Gender"])
gender2groups_df.count().head()
#1. Unique Items (unique_tot)
gen2_counts = gender2groups_df["Gender"].count()
#gen2_counts
#----------------------------
#Gcounts
#purchase_data["Item ID"].unique()
#gen_unique_tot = gender_groups["Item ID"].count()
#gen_unique_df = gen_unique_tot.to_frame().reset_index()
#gen_unique_df.head()
#purchase_gender = gen_df["Item ID"].count()
#purchase_df.head()
#pur_data1["Number of Unique Items"] = purchase_data["Item ID"].max()
#-------------------------------
#2. Average Price (average_price)
gen_average_price = gender2groups_df["Price"].mean()
#gen_average_price_format = f'${gen_average_price:.2f}'
#average_price
#-------------------------------
#gen_unique_df["Average Purchase Price"] = purchase_df["Price"].mean()
#gen_unique_df
#average_price_format = f'${average_price:.2f}'
#-------------------------------
#4. Total Revenue (rev_tot)
gen_tot_rev = gender2groups_df["Price"].sum()
#gen_tot_rev_format = f'${gen_tot_rev:.2f}'
#rev_tot_format = f'${average_price:.2f}'
#----------
#5  Average Total Purchase per Person (avgpp)
#avgpp = (gen_rev_tot) / gender_groups['SN'].nunique()
avgpp = gen_tot_rev / gender_df
#avgpp_format = f'${avgpp:.2f}'
#avgpp
gen_totals_df = pd.DataFrame({"Purchase Counts": gen2_counts, "Average Purchase Price":gen_average_price, 
                         "Total Purchase Value": gen_tot_rev, "Avg Total Purchase per Person": avgpp})
gen_totals_df["Average Purchase Price"] = gen_totals_df["Average Purchase Price"].map("${:.2f}".format)
gen_totals_df["Total Purchase Value"] = gen_totals_df["Total Purchase Value"].map("${:,.2f}".format)
gen_totals_df["Avg Total Purchase per Person"] = gen_totals_df["Avg Total Purchase per Person"].map("${:.2f}".format)
gen_totals_df.head()

Purchase Counts Average Purchase Price  \
Gender                                                          
Female                             113                  $3.20   
Male                               652                  $3.02   
Other / Non-Disclosed               15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [12]:
#print(purchase_data["Age"].max())
#print(purchase_data["Age"].min())

# Create the bins in which Data will be held
# Bins are 0, 60, 70, 80, 90, 100
bins = [0, 9, 14, 19, 24, 29, 34, 39, 46]
#bins = [0, 10, 15, 20, 25, 30, 35, 40, 45]
#bins = [0, 11, 16, 21, 26, 31, 36, 41, 45]

# Create the names for the four bins
group_labels = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
purchase_data["Age_Bin"] = pd.cut(purchase_data["Age"], bins, right=True, labels=group_labels, include_lowest=True)
purchase_data.head()

# Create a GroupBy object based upon "View Group"
#age_count = purchase_data['SN'].count()
age_group = purchase_data.groupby("Age_Bin")
age_count = age_group['SN'].count()

age_group_df = age_count.to_frame().reset_index()
#age_group_df

agesum = age_group_df["SN"].sum()
#agesum

age_group_df["Percentage of Players"] = (age_group_df["SN"] / agesum) *100
age_group_df["Percentage of Players"] = age_group_df["Percentage of Players"].map("{:.2f}".format)
#gender2_sort_df = gender2_df.sort_values("Percentage of Players", ascending = False)
age_group_df = age_group_df.rename(columns={"SN":"Total Counts"})
#gender2_sort_df.head()
age_group_df

Age_Bin  Total Counts Percentage of Players
0     <10            23                  2.95
1   10-14            28                  3.59
2   15-19           136                 17.44
3   20-24           365                 46.79
4   25-29           101                 12.95
5   30-34            73                  9.36
6   35-39            41                  5.26
7     40+            13                  1.67

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [13]:
purchase_data.head()
#age_counts = purchase_data["Age_Bin"].value_counts()
#age_counts
#1. Unique Items (unique_tot)
age2groups_df = purchase_data.groupby(["Age_Bin"])
age_counts = age2groups_df["Age_Bin"].count()
#age_counts
#----------------------------
#2. Average Price (average_price)
age_average_price = age2groups_df["Price"].mean()
#gen_average_price_format = f'${gen_average_price:.2f}'
#average_price
#-------------------------------
#3. Total Revenue (rev_tot)
age_tot_rev = age2groups_df["Price"].sum()
#gen_tot_rev_format = f'${gen_tot_rev:.2f}'
#rev_tot_format = f'${average_price:.2f}'
#----------
#4.  Average Total Purchase per Person (avgpp)
#avgpp = (gen_rev_tot) / gender_groups['SN'].nunique()
agecnt_df = age_group['SN'].nunique()
ageavgpp = age_tot_rev / agecnt_df
#avgpp_format = f'${avgpp:.2f}'
#avgpp
age_totals_df = pd.DataFrame({"Purchase Counts": age_counts, "Average Purchase Price":age_average_price, 
                         "Total Purchase Value": age_tot_rev, "Avg Total Purchase per Person": ageavgpp})
age_totals_df["Average Purchase Price"] = age_totals_df["Average Purchase Price"].map("${:.2f}".format)
age_totals_df["Total Purchase Value"] = age_totals_df["Total Purchase Value"].map("${:,.2f}".format)
age_totals_df["Avg Total Purchase per Person"] = age_totals_df["Avg Total Purchase per Person"].map("${:.2f}".format)
age_totals_df

Purchase Counts Average Purchase Price Total Purchase Value  \
Age_Bin                                                                
<10                   23                  $3.35               $77.13   
10-14                 28                  $2.96               $82.78   
15-19                136                  $3.04              $412.89   
20-24                365                  $3.05            $1,114.06   
25-29                101                  $2.90              $293.00   
30-34                 73                  $2.93              $214.00   
35-39                 41                  $3.60              $147.67   
40+                   13                  $2.94               $38.24   

        Avg Total Purchase per Person  
Age_Bin                                
<10                             $4.54  
10-14                           $3.76  
15-19                           $3.86  
20-24                           $4.32  
25-29                           $3.81  
30-34                           $4.12  
35-39                           $4.76  
40+                             $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [14]:
purchase_data.head()
sngroups_df = purchase_data.groupby(["SN"])
sn_counts = sngroups_df["Gender"].count()
#sn_counts
#----------------------------
#2. Average Price (average_price)
sn_average_price = sngroups_df["Price"].mean()
#gen_average_price_format = f'${gen_average_price:.2f}'
#average_price
#-------------------------------
#3. Total Revenue (rev_tot)
sn_tot_rev = sngroups_df["Price"].sum()
#gen_tot_rev_format = f'${gen_tot_rev:.2f}'
#rev_tot_format = f'${average_price:.2f}'
#----------
#4.  Average Total Purchase per Person (avgpp)
#avgpp = (gen_rev_tot) / gender_groups['SN'].nunique()
#sncnt_df = sngroups_df['SN'].nunique()
#snavgpp = sn_tot_rev / sncnt_df
#avgpp_format = f'${avgpp:.2f}'
#avgpp
sn_totals_df = pd.DataFrame({"Purchase Counts": sn_counts, "Average Purchase Price":sn_average_price, 
                         "Total Purchase Value": sn_tot_rev})
sn2_sort_df = sn_totals_df.sort_values("Total Purchase Value", ascending = False)
sn2_sort_df["Average Purchase Price"] = sn2_sort_df["Average Purchase Price"].map("${:.2f}".format)
sn2_sort_df["Total Purchase Value"] = sn2_sort_df["Total Purchase Value"].map("${:,.2f}".format)
#sn2_sort_df["Avg Total Purchase per Person"] = sn2_sort_df["Avg Total Purchase per Person"].map("${:.2f}".format)

sn2_sort_df.head()
#sn_totals_df.head()

Purchase Counts Average Purchase Price Total Purchase Value
SN                                                                      
Lisosia93                  5                  $3.79               $18.96
Idastidru52                4                  $3.86               $15.45
Chamjask73                 3                  $4.61               $13.83
Iral74                     4                  $3.40               $13.62
Iskadarya95                3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [15]:
purchase_data
item_groups_df = purchase_data.groupby(["Item ID", "Item Name"])
# 1. Total Count
item_counts = item_groups_df["Gender"].count()
#2. Average Price (average_price)
#item_average_price = sngroups_df["Price"].mean()
#3. Total Revenue (rev_tot)
item_tot_rev = item_groups_df["Price"].sum()
#4.  Average Total Purchase per Person (avgpp)
item_price = item_tot_rev / item_counts
#avgpp = (gen_rev_tot) / gender_groups['SN'].nunique()
#itemcnt_df = sngroups_df['SN'].nunique()
#itemavgpp = sn_tot_rev / sncnt_df
#avgpp_format = f'${avgpp:.2f}'
#avgpp
item_totals_df = pd.DataFrame({"Purchase Counts": item_counts, "Item Price": item_price, 
                         "Total Purchase Value": item_tot_rev})
item2_sort_df = item_totals_df.sort_values("Purchase Counts", ascending = False)
item2_sort_df["Item Price"] = item2_sort_df["Item Price"].map("${:.2f}".format)
item2_sort_df["Total Purchase Value"] = item2_sort_df["Total Purchase Value"].map("${:.2f}".format)
item2_sort_df.head()


Purchase Counts  \
Item ID Item Name                                                       
178     Oathbreaker, Last Hope of the Breaking Storm               12   
145     Fiery Glass Crusader                                        9   
108     Extraction, Quickblade Of Trembling Hands                   9   
82      Nirvana                                                     9   
19      Pursuit, Cudgel of Necromancy                               8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [16]:
item3_sort_df = item_totals_df.sort_values("Total Purchase Value", ascending = False)
item3_sort_df["Item Price"] = item3_sort_df["Item Price"].map("${:.2f}".format)
item3_sort_df["Total Purchase Value"] = item3_sort_df["Total Purchase Value"].map("${:.2f}".format)
item3_sort_df.head()

Purchase Counts  \
Item ID Item Name                                                       
178     Oathbreaker, Last Hope of the Breaking Storm               12   
82      Nirvana                                                     9   
145     Fiery Glass Crusader                                        9   
92      Final Critic                                                8   
103     Singed Scalpel                                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

## Conclusions

In evaluating the various analysis on the Pymoli data, noted the following 3 items:

* The most profitable and most popular item is 178, Oathbreaker, Last Hope of the Breaking Storm, with 4 of the main items bringing in much more than any of the other items.  Need to identify what in those items drives the popularity and generate more of these types of items.

* The original noted conclusions were accurate, that a vast majority are male (84%) and a smaller, but notable proportion of female players (14%).  The purchase per female of 4.47 does appear to be higher than males at 4.07 on average.  Those that didn't identify themselves brought in the most per user at 4.56 on average.

* The age group of 15-30 appears to drive a majority of the spending and traffic.  The average user in this age demographic had an average purchase price of 3.05, where the higher average purchase prices was under age 10 and age 30-35 running between 3.30 and 3.60; with lower purchase volumes.
